Search: Remove unnecessary whitespace in case if we remove part of the string

vovakulikov opened the pull request:
Fixes https://github.com/sourcegraph/sourcegraph/issues/41452

This PR removes unnecessary in case if we remove something from the query. We do it in a few cases. For example when we extract from the query filters that are supposed to be a part of UI (pattern type and case: filters)

**Prior to this PR:** 
 on the homepage we have `insights case:yes repo: sourcegraph/sourcegraph`
 on the search result page we have `insights  repo:sourcegraph/sourcegraph` (note that between search pattern `insights` and repo filter **there is an additional space**)
 
**In this PR** 
 on the homepage we have `insights case:yes repo: sourcegraph/sourcegraph`
 on the search result page we have `insights repo:sourcegraph/sourcegraph` Note that **there is only one space** between `insights` and repo filters

## Test plan
- Search any querh that has filters that are supposed to be remove from the query and become part of UI controls (like pattern type or case: filters) for example `insights case:yes patterntype:regex repo:sourcegraph/sourcegraph`
- On the search result page you should see that there are no whitespaces between parts that are still in the query (like `insights repo:sourcegraph/sourcegraph`, note that between search pattern and repo filter is no spaces) 
- Click different fitlers in the search sidebar panel like search types (find a symbol or search diffs), make sure that query is not currapted and still is valid after changing it through those filters. 


<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

## App preview:

- [Web](https://sg-web-vk-fix-replace-range.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-chpkmfarzg.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

rvantonder reviewed (commented):
Can we add a test for this behavioral change?

As a side note, `replaceRange`/substring replacement is not a great way to manipulate the query. I realize it kind of exists in some places in the code right now so it's easier to try fix this function, but it would be a better to use something like [`stringHuman`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@2d66e9c5b9bcfaca322acd89e2eb07d8d25f4b1a/-/blob/client/web/src/enterprise/insights/pages/insights/creation/search-insight/utils/use-url-query-insight/use-url-query-insight.ts?L55) so query strings are formatted and correct by construction

vovakulikov commented:
@rvantonder, thanks for the comment. 
>As a side note, replaceRange/substring replacement is not a great way to manipulate the query

I had the same feeling while I was changing this function. This logic seems too business-specific to be placed in such a low-level API as `replaceRange`. 

I don't feel comfortable changing `stringHuman` function, though. It seems that our problem is only related to the query manipulation and not the query formatting. We should have another level of API for query manipulation like `replaceFilter` for operation where we replace/remove filters from the query and consider such specific peculiarities as removing or preserving whitespaces.

unclejustin commented:
Approved, but I still don't like it this close to code freeze. I recommend we get this merged as soon as possible so that we have a higher chance of finding any problems.

coury-clark reviewed (changes requested):
Agreed, some test cases to prove this behavior before merging please!

vovakulikov commented:
@rvantonder @coury-clark I changed the fix. Instead of patching `replaceRange`, I found that we've been using it widely for deleting filters from queries, but it turned out that we already have a proper transformer for it - `omitFilters`. I fixed this transformer instead and replaced all usage of `replaceRange`, which was used as a delete transformer. (it seems that `replaceRange` usages were just an implementation leak in most consumers, from my perspective we should not expose such low level API and we should use it only in transformers)

@coury-clark `omitFilters` already have unit tests. I fixed them since the whitespace trim logic now works as expected.

vovakulikov commented on client/shared/src/util/url.ts:
```diff
@@ -538,17 +537,16 @@ export function buildSearchURLQuery(
     if (globalPatternType?.value) {
-        const { start, end } = globalPatternType.range
         patternTypeParameter = globalPatternType.value.value
-        queryParameter = replaceRange(queryParameter, { start: Math.max(0, start - 1), end }).trim()
```
@rvantonder I think it doesn't break the flow here but just double check - do you remember why did you need to subtract here one position from the start here?

rvantonder replied:
Introduced before my time, so I don't know: https://github.com/sourcegraph/sourcegraph/pull/10609/files#diff-c46ab76b44d7a7905b4e75d4a7acf08bf3315504531d3305758a26f783edd448R586

There's probably a fairly heavy dependency/test on this function so I think we would notice if it broke.
